{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Annotating Human Proteome query results containing site and region features onto posID level data\n",
    "3.22.2022 \n",
    "@mfpfox\n",
    "\n",
    "### help links\n",
    "* Sequence annotation (Features) <https://www.uniprot.org/help/sequence_annotation>\n",
    "* api query <https://www.uniprot.org/help/api_queries>\n",
    "\n",
    "---\n",
    "\n",
    "# **columns parsed/formatted by downloadUniprotDomains.py:**\n",
    "## SITES\n",
    "1. SubstrateBindingSite\n",
    "2. ActiveSite\n",
    "3. MetalBindSite\n",
    "4. ModifiedResidue\n",
    "5. Lipidation\n",
    "6. Glycosylation\n",
    "7. Mutagenesis\n",
    "8. DisulfideBond\n",
    "\n",
    "FORMAT: seperator \"; \" disulfide sep '..'\n",
    "\n",
    "## REGIONS\n",
    "0. topological domain (not included in saved output from script)\n",
    "1. Domain\n",
    "2. NucleotideBindRegion\n",
    "3. DNAbindRegion\n",
    "4. ZincFinger\n",
    "5. Motif\n",
    "6. DisorderedRegion\n",
    "7. BetaStrand\n",
    "8. Helix\n",
    "9. Turn\n",
    "\n",
    "FORMAT: regions '-' and '; ' \n",
    "\n",
    "## proximal defined by not at site and not at region upper or lower bound position, but within +5 of high and -5 positions of region bounds and +- 5 of site positions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>.container {width:90% !important;}</style>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import sys\n",
    "import os\n",
    "sys.path.append(\"/Users/mariapalafox/Desktop/TOOLBOXPY\")\n",
    "from all_funx import *\n",
    "from IPython.display import display, HTML\n",
    "from IPython.display import Image\n",
    "display(HTML(\"<style>.container {width:90% !important;}</style>\"))\n",
    "pd.set_option('display.max_columns', None)\n",
    "pd.set_option('display.max_rows', 40)\n",
    "pd.set_option('display.max_colwidth', 1000)\n",
    "pd.options.display.max_seq_items = 2000\n",
    "os.chdir(\"/Users/mariapalafox/Desktop/BRIDGE/P4_UniProt_Query/isPositionROI/\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Filtered downloadUniprotDomains.py output file that was QC'd for ukb sequence match\n",
    "\n",
    "* dir: `/Users/mariapalafox/Desktop/BRIDGE/P4_UniProt_Query/isPositionROI/Uniprot_Domains/20220212/`\n",
    "\n",
    "* unfiltered file made by downloadUniprotDomains.py:`features_summary_QCd_sequences.csv`\n",
    "\n",
    "### Proceeding to annotated ukb sites and regions (exact match OR proximal match) for positions of sequences that have identical UKB 2/12/22 & UKB CpDAA ref sequences\n",
    "\n",
    "* filtered ukb annotation data made in above cells:`features_summary_QCd_IdenticalSeq_18231proteins.csv`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of posID proteins found in ukb annotation df \n",
      " 38 \n",
      " n=4497 now, was 4535 pre merge\n",
      " Percent protein IDs lost: 0.8379272326350606\n",
      "Total of  119  CpDAA posIDs lost from merge\n",
      "had 18827, now have 18708 cpdaa\n"
     ]
    }
   ],
   "source": [
    "print(\"Number of posID proteins found in ukb annotation df \\n\",\n",
    "      4535-4497, \"\\n\", \"n=4497 now, was 4535 pre merge\\n\",\n",
    "      \"Percent protein IDs lost:\" ,(38/4535)*100)\n",
    "print('Total of ', 18827-18708, ' CpDAA posIDs lost from merge')\n",
    "print(\"had 18827, now have 18708 cpdaa\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import ukb annotation file drop sequence col reorder cols\n",
    "\n",
    "dirname = \"Uniprot_Domains/20220212/\"\n",
    "df = pd.read_csv(dirname+\"features_summary_QCd_sequences.csv\")\n",
    "#--------------------------------------------\n",
    "# save protein level file\n",
    "#  # protein -level merge later\n",
    "# 'Protein', 'Gene', 'Length','SubunitStructure', '3D', 'PDB', 'InteractsWith', 'PTM', \n",
    "proteinlevel = df[['UKBID', \n",
    "         'identicalToRefSequence',\n",
    "           'query.minus.REF.difference', \n",
    "             'QC.posID.difference',\n",
    "                 'Protein', 'Gene', 'Length','SubunitStructure',\n",
    "               '3D', 'PDB', 'InteractsWith', 'PTM']].copy()\n",
    "proteinlevel.to_csv(\"features_summary_QCd_02122022_proteinLevelColumns.csv\", index=False)\n",
    "#--------------------------------------------\n",
    "# simplifly  'Sequence', \n",
    "df = df[['UKBID', \n",
    "         'identicalToRefSequence',\n",
    "           'query.minus.REF.difference', \n",
    "             'QC.posID.difference',\n",
    "                 'Protein', 'Gene', 'Length','SubunitStructure',\n",
    "               '3D', 'PDB', 'InteractsWith', 'PTM', \n",
    "             'SubstrateBindingSite',\n",
    "             'ActiveSite', \n",
    "             'MetalBindSite', \n",
    "             'ModifiedResidue',\n",
    "             'Lipidation',\n",
    "             'Glycosylation', \n",
    "             'Mutagenesis', \n",
    "                'DisulfideBond',\n",
    "    'NucleotideBindRegion', 'DNAbindRegion',\n",
    "   'ZincFinger', 'Motif', \n",
    "         'DisorderedRegion', 'Domain', \n",
    "   'BetaStrand', 'Helix', \n",
    "         'Turn'\n",
    "         ]].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   identicalToRefSequence  Count\n",
      "0                    True  18231\n",
      "1                   False    185\n"
     ]
    }
   ],
   "source": [
    "#qc\n",
    "print(checkColumnValues(df, 'identicalToRefSequence'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(185, 29)\n"
     ]
    }
   ],
   "source": [
    "# save dropped protein annotation rows \n",
    "falsedf =  df[~df['identicalToRefSequence']].copy()\n",
    "print(falsedf.shape)\n",
    "\n",
    "falsedf.to_csv(\"features_summary_QCd_NotIdenticalSequence_185proteins.csv\", index=False)\n",
    "# saving falsedf , sequences dropped because they differed from the reference proteome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(18231, 29)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UKBID</th>\n",
       "      <th>identicalToRefSequence</th>\n",
       "      <th>query.minus.REF.difference</th>\n",
       "      <th>QC.posID.difference</th>\n",
       "      <th>Protein</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Length</th>\n",
       "      <th>SubunitStructure</th>\n",
       "      <th>3D</th>\n",
       "      <th>PDB</th>\n",
       "      <th>InteractsWith</th>\n",
       "      <th>PTM</th>\n",
       "      <th>SubstrateBindingSite</th>\n",
       "      <th>ActiveSite</th>\n",
       "      <th>MetalBindSite</th>\n",
       "      <th>ModifiedResidue</th>\n",
       "      <th>Lipidation</th>\n",
       "      <th>Glycosylation</th>\n",
       "      <th>Mutagenesis</th>\n",
       "      <th>DisulfideBond</th>\n",
       "      <th>NucleotideBindRegion</th>\n",
       "      <th>DNAbindRegion</th>\n",
       "      <th>ZincFinger</th>\n",
       "      <th>Motif</th>\n",
       "      <th>DisorderedRegion</th>\n",
       "      <th>Domain</th>\n",
       "      <th>BetaStrand</th>\n",
       "      <th>Helix</th>\n",
       "      <th>Turn</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>Q9GZX5</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Zinc finger protein 350 (KRAB zinc finger protein ZFQR) (Zinc finger and BRCA1-interacting protein with a KRAB domain 1) (Zinc finger protein ZBRK1)</td>\n",
       "      <td>ZNF350</td>\n",
       "      <td>532</td>\n",
       "      <td>SUBUNIT: Interacts with BRCA1. Interacts with RNF11. {ECO:0000269|PubMed:11090615, ECO:0000269|PubMed:14755250}.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P38398</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>206-228; 234-256; 262-284; 290-312; 318-340; 346-368; 374-396; 402-424</td>\n",
       "      <td>NaN</td>\n",
       "      <td>427-465</td>\n",
       "      <td>8-79</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    UKBID  identicalToRefSequence  query.minus.REF.difference  \\\n",
       "0  Q9GZX5                    True                           0   \n",
       "\n",
       "  QC.posID.difference  \\\n",
       "0                 NaN   \n",
       "\n",
       "                                                                                                                                                Protein  \\\n",
       "0  Zinc finger protein 350 (KRAB zinc finger protein ZFQR) (Zinc finger and BRCA1-interacting protein with a KRAB domain 1) (Zinc finger protein ZBRK1)   \n",
       "\n",
       "     Gene  Length  \\\n",
       "0  ZNF350     532   \n",
       "\n",
       "                                                                                                   SubunitStructure  \\\n",
       "0  SUBUNIT: Interacts with BRCA1. Interacts with RNF11. {ECO:0000269|PubMed:11090615, ECO:0000269|PubMed:14755250}.   \n",
       "\n",
       "    3D  PDB InteractsWith  PTM SubstrateBindingSite ActiveSite MetalBindSite  \\\n",
       "0  NaN  NaN        P38398  NaN                  NaN        NaN           NaN   \n",
       "\n",
       "  ModifiedResidue Lipidation Glycosylation Mutagenesis DisulfideBond  \\\n",
       "0             NaN        NaN           NaN         NaN           NaN   \n",
       "\n",
       "  NucleotideBindRegion DNAbindRegion  \\\n",
       "0                  NaN           NaN   \n",
       "\n",
       "                                                               ZincFinger  \\\n",
       "0  206-228; 234-256; 262-284; 290-312; 318-340; 346-368; 374-396; 402-424   \n",
       "\n",
       "  Motif DisorderedRegion Domain BetaStrand Helix Turn  \n",
       "0   NaN          427-465   8-79        NaN   NaN  NaN  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# drop rows with different sequences between query and ref for cpdaa\n",
    "df =  df[df['identicalToRefSequence']].copy()\n",
    "print(df.shape)\n",
    "df.head(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "# saving\n",
    "df.to_csv(\"features_summary_QCd_IdenticalSeq_18231proteins.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# isposition_atSite_inDomain.py functions for SITE and REGION ukb annotations\n",
    "\n",
    "##### UPDATE THOUGHTS isposition_atSite_inDomain.py :\n",
    "- add section of code that checks if query result sequence is identical to ref of cpdaa and category positions\n",
    "    - if sequence is not identical, are positions of cpdaa and category unchanged?\n",
    "    - if sequence is same length but different, qc that site modifications match reference aa type at position\n",
    "- **add substrings corresponding to zinc fingers, domains, other regions that may be of interest for aa abundance analysis**\n",
    "\n",
    "### tested functions in markdown, final functions in script\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```python\n",
    "#------------------------------code for script------------------------\n",
    "#-------variables--------------\n",
    "\n",
    "# 2 ARGS for filenames,all positions MendelCpD: codonMaxMeanScored_mergedDescribeProt_1231_MendelCpD.csv\n",
    "POSINFILE = \"REF_posID_level_18827cpdaa_4535UKBIDs.csv\"\n",
    "# import ukb annotation reference with only identical ukb sequences\n",
    "UKBINFILE = \"features_summary_QCd_IdenticalSeq_18231proteins.csv\"\n",
    "# 2 ARGS name of column with UKBIDs for both dataframes\n",
    "IDname = \"UKBID\"\n",
    "posIDname = \"posID\"\n",
    "saveSite = \"CpDAA_UKBfeatures_Sites_mergedAnnotations.csv\"\n",
    "saveRegion = \"CpDAA_UKBfeatures_Region_mergedAnnotations.csv\"\n",
    "#-------what code does--------------\n",
    "# imports and merges on position index ref ukbIDs, so only looking at proteins with annotations from ukb\n",
    "# adds an integer pos column from split posID\n",
    "# splits merged df into site and region dfs, drops columns other than posID, pos, site|regoin annotions\n",
    "# *makes new header column with exact and proximal colnames added for site and region dfs\n",
    "\n",
    "positions = pd.read_csv(POSINFILE)\n",
    "ukb = pd.read_csv(UKBINFILE)\n",
    "merged1 = pd.merge(positions, ukb, on=[IDname], how=\"inner\") # MERGES UKB ANNOTATION ONTO POSITION LEVEL DF\n",
    "merged1= getAAposition(merged1, 'posID', 'pos') # ADDS POS INTEGER COL BY SPLITTING POSID\n",
    "\n",
    "# *MAKE HEADER COLUMNS FOR MERGED DF'S SPECIFIC TO SITE AND REGION COLUMNS\n",
    "basicHead = ['posID', 'pos']\n",
    "sitesCol= ['SubstrateBindingSite', 'ActiveSite', 'MetalBindSite',\n",
    "       'ModifiedResidue', 'Lipidation', 'Glycosylation', 'Mutagenesis',\n",
    "       'DisulfideBond'] # SITE\n",
    "sitesDF = merged1[basicHead + sitesCol].copy()\n",
    "sitesDF.to_csv(saveSite, index=False) # save site and region df's\n",
    "sitesAddCol = [] # make exact and proximal header additions SITES\n",
    "for coli in sitesCol:\n",
    "    exx = coli + '_Exact'\n",
    "    prx = coli + '_Proximal5'\n",
    "    sitesAddCol.append(coli)\n",
    "    sitesAddCol.append(exx)\n",
    "    sitesAddCol.append(prx)\n",
    "regionCol = [\n",
    "        'NucleotideBindRegion', 'DNAbindRegion', 'ZincFinger',\n",
    "       'Motif', 'DisorderedRegion', 'Domain', 'BetaStrand', 'Helix', \n",
    "        'Turn'] # REGION\n",
    "regionDF = merged1[basicHead + regionCol].copy()\n",
    "regionDF.to_csv(saveRegion, index=False) # save site and region df's\n",
    "regAddCol = [] # make exact and proximal header additions REGIONS\n",
    "for coli in regionCol:\n",
    "    exx = coli + '_Exact'\n",
    "    prx = coli + '_Proximal5'\n",
    "    regAddCol.append(coli)\n",
    "    regAddCol.append(exx)\n",
    "    regAddCol.append(prx)\n",
    "# NEW HEADERS\n",
    "newheaderSites = basicHead + sitesAddCol\n",
    "newheaderRegion = basicHead + regAddCol\n",
    "\n",
    "# Function call outfile naming for site and region\n",
    "outSite = saveSite.replace(\"_mergedAnnotations.csv\", \"_ANNOTATED.csv\")\n",
    "outRegion = saveRegion.replace(\"_mergedAnnotations.csv\", \"_ANNOTATED.csv\")\n",
    "\n",
    "addSiteAnnotations(saveSite, newheaderSites, outSite, sitesCol)\n",
    "addRegionAnnotations(saveRegion, newheaderRegion, outRegion, regionCol)\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```python\n",
    "def addSiteAnnotations(infile, head, outfile, siteCol):\n",
    "    f = open(infile)\n",
    "    df = list(csv.reader(f))\n",
    "    # df = df[0:2]\n",
    "    with open(outfile, 'w') as out:\n",
    "            csvWriter = csv.writer(out)\n",
    "            csvWriter.writerow(head)\n",
    "    # loop over rows skip header\n",
    "    for row in df[1:]:\n",
    "        #print(row)\n",
    "        newrow = []\n",
    "        newrow.append(row[0]) # posID\n",
    "        pos = int(row[1])\n",
    "        newrow.append(pos)\n",
    "        annoRow = row[2:] # select site annotation columns, skip posID and pos\n",
    "        for i, v in enumerate(annoRow):\n",
    "            flounder = False # FLAG EXACT\n",
    "            flounder2 = False # FLAG PROXIMAL\n",
    "            siteType = siteCol[i]\n",
    "            # nan annotation value case\n",
    "            if (len(v) < 1) | (v == ''): \n",
    "                newrow.append(v) \n",
    "                newrow.append(flounder) # exact \n",
    "                newrow.append(flounder2) # proximal\n",
    "            # not nan annotation value case\n",
    "            if len(v) >= 1: \n",
    "                if siteType != 'DisulfideBond':\n",
    "                    vsplit = v.split(\";\")\n",
    "                    # EXACT\n",
    "                    for exac in vsplit:\n",
    "                        try:\n",
    "                            posCheck = int(exac.strip(\" \"))\n",
    "                        except:\n",
    "                            print(\"error posCheck:\", posCheck)\n",
    "                            posCheck = -100\n",
    "                        # check exact match \n",
    "                        if pos == posCheck:\n",
    "                            flounder = True\n",
    "                            break\n",
    "                    # PROXIMAL\n",
    "                    for prox in vsplit:\n",
    "                        try:\n",
    "                            posCheck = int(prox.strip(\" \"))\n",
    "                        except:\n",
    "                            print(\"error posCheck:\", posCheck)\n",
    "                            posCheck = -100\n",
    "                        # check proximal match\n",
    "                        if (pos >= posCheck-5 and pos < posCheck) | (pos <= posCheck+5 and pos > posCheck):\n",
    "                            flounder2 = True\n",
    "                            break\n",
    "                    newrow.append(v)\n",
    "                    newrow.append(flounder)\n",
    "                    newrow.append(flounder2)\n",
    "                # disulfide case          \n",
    "                if siteType == 'DisulfideBond':\n",
    "                    vsplit = v.split(\";\")\n",
    "                    for exac in vsplit:\n",
    "                        posCheck = exac.strip(\" \")\n",
    "                        posCheck = posCheck.split(\"..\")\n",
    "                        for disul in posCheck:\n",
    "                            try:\n",
    "                                di = int(disul)\n",
    "                            except: \n",
    "                                print(\"error disul:\", disul)\n",
    "                                di = -100\n",
    "                            # check exact match\n",
    "                            if pos == di:\n",
    "                                flounder = True\n",
    "                                break\n",
    "                    for prox in vsplit:\n",
    "                        posCheck = prox.strip(\" \")\n",
    "                        posCheck = posCheck.split(\"..\")\n",
    "                        for pp in posCheck:\n",
    "                            try:\n",
    "                                pc = int(pp)\n",
    "                            except:\n",
    "                                print(\"error pp:\", pp)\n",
    "                                pc = -100\n",
    "                            # check proximal match\n",
    "                            if (pos >= pc-5 and pos < pc) | (pos <= pc+5 and pos > pc):\n",
    "                                flounder2 = True\n",
    "                                break\n",
    "                    newrow.append(v)\n",
    "                    newrow.append(flounder)\n",
    "                    newrow.append(flounder2)\n",
    "        with open(outfile, 'a') as out:\n",
    "            csvWriter = csv.writer(out)\n",
    "            csvWriter.writerow(newrow)\n",
    "#addSiteAnnotations(saveSite, newheaderSites, outSite, sitesCol)\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```python\n",
    "def addRegionAnnotations(infile, head, outfile, siteCol):\n",
    "    f = open(infile)\n",
    "    df = list(csv.reader(f))\n",
    "    #df = df[0:2]\n",
    "    with open(outfile, 'w') as out:\n",
    "            csvWriter = csv.writer(out)\n",
    "            csvWriter.writerow(head)\n",
    "    # loop over rows skip header\n",
    "    for row in df[1:]:\n",
    "        newrow = []\n",
    "        newrow.append(row[0]) # posID\n",
    "        pos = int(row[1]) # aa position to check in domain range\n",
    "        newrow.append(pos)\n",
    "        annoRow = row[2:] # select region annotation columns, skip posID and pos cols\n",
    "        for i, v in enumerate(annoRow):\n",
    "            flounder = False # FLAG EXACT\n",
    "            flounder2 = False # FLAG PROXIMAL\n",
    "            siteType = siteCol[i]\n",
    "            # nan annotation value case, ADD EMPTY STRING TO FILTER\n",
    "            if (len(v) < 1) | (v == ''): \n",
    "                newrow.append(v) \n",
    "                newrow.append(flounder) # exact \n",
    "                newrow.append(flounder2) # proximal\n",
    "            # not nan annotation value case\n",
    "            if len(v) >= 1:     \n",
    "                vsplit = v.split(\";\")\n",
    "                # EXACT\n",
    "                for exac in vsplit:\n",
    "                    posCheck = exac.strip(\" \")\n",
    "                    posCheck = posCheck.split(\"-\")\n",
    "                    if len(posCheck) == 2:\n",
    "                        b1 = int(posCheck[0])\n",
    "                        b2 = int(posCheck[1])\n",
    "                        r = [*range(b1, b2+1, 1)] # include 22,24 is 22,23,24 range\n",
    "                        if pos in r:\n",
    "                            # check exact match \n",
    "                            flounder = True\n",
    "                            break      \n",
    "                    if len(posCheck) == 1:\n",
    "                        try:\n",
    "                            b1 = int(posCheck[0])\n",
    "                        except:\n",
    "                            print(\"error posCheck:\", posCheck)\n",
    "                            b1=-100\n",
    "                        if pos == b1:\n",
    "                            # check exact match \n",
    "                            flounder = True\n",
    "                            break     \n",
    "                # PROXIMAL\n",
    "                for prox in vsplit:\n",
    "                    posCheck = prox.strip(\" \")\n",
    "                    posCheck = posCheck.split(\"-\")\n",
    "                    try:                            \n",
    "                        b1 = int(posCheck[0])\n",
    "                        b2 = int(posCheck[1])\n",
    "                    except:\n",
    "                        print(\"error posCheck:\", posCheck)\n",
    "                        b1=-100\n",
    "                        b2=-100\n",
    "                    # check proximal match, within |5 aa| from lower and upper bound\n",
    "                    if (pos >= b1-5 and pos < b1) | (pos <= b2+5 and pos > b2):\n",
    "                        flounder2 = True\n",
    "                        break\n",
    "                newrow.append(v)\n",
    "                newrow.append(flounder)\n",
    "                newrow.append(flounder2)                    \n",
    "        with open(outfile, 'a') as out:\n",
    "            csvWriter = csv.writer(out)\n",
    "            csvWriter.writerow(newrow)\n",
    "            \n",
    "#addRegionAnnotations(saveRegion, newheaderRegion, outRegion, regionCol)\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['22', '24']\n",
      "[22, 23, 24]\n"
     ]
    }
   ],
   "source": [
    "# qc\n",
    "l= ['22-24']\n",
    "for pp in l:\n",
    "    pc = pp.strip(' ')\n",
    "    pc = pc.split('-')\n",
    "    print(pc)\n",
    "    b1 = int(pc[0])\n",
    "    b2 = int(pc[1])\n",
    "    r = [*range(b1, b2+1, 1)] # inclusive\n",
    "    print(r)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
